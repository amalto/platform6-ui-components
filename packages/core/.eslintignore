dist/
loader/

src/components.d.ts