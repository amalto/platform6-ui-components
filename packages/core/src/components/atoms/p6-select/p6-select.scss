$choices-border-radius: $control-border-width;
$choices-border-radius: $input-radius;
$choices-border-radius-item: 0;
$choices-bg-color: $input-background-color;
$choices-bg-color-disabled: $input-disabled-background-color;
$choices-bg-color-dropdown: $input-background-color;
$choices-text-color: $input-color;
$choices-keyline-color: $input-border-color;
$choices-primary-color: $grey-lightest;

$p6-select-border-color: #e8ebed !default;
$p6-select-border-width: 1px !default;
$p6-select-multiple-item-bg-color: #e8ebed !default;
$p6-select-multiple-item-text-color: #555 !default;
$p6-select-dropdown-item-highlighted-bg-color: #dce7fe !default;

@import '~choices.js/src/styles/choices';
@import '../../../styles/atoms/control';

@mixin cross($color) {
  text-indent: -9999px;
  appearance: none;
  border: 0;
  background: initial;
  position: absolute;
  opacity: 0.3;
  top: calc(50% - 0.125em);
  height: 0.625em;
  width: 0.625em;
  margin: -0.437em 0 0;

  &:hover {
    opacity: 1;
  }
  &::before,
  &::after {
    position: absolute;
    left: 0.5em;
    content: ' ';
    height: calc(1em + 1px);
    width: 2px;
    background-color: $color;
  }
  &::before {
    transform: rotate(45deg);
  }
  &::after {
    transform: rotate(-45deg);
  }
}

$sizes: ('default' $size-small, small $size-small, normal $size-normal, medium $size-medium, large $size-large);

p6-select {
  // Sizes
  @each $list in $sizes {
    $name: nth($list, 1);
    $size: nth($list, 2);
    &.is-#{$name} {
      .#{$choices-selector},
      .#{$choices-selector}__inner,
      .#{$choices-selector}__item,
      .#{$choices-selector}__input,
      .p6__#{$choices-selector}__button,
      .#{$choices-selector}[data-type*='select-multiple'] .#{$choices-selector}__item {
        font-size: $size;
      }
    }
  }

  // Colors
  @each $name, $pair in $colors {
    $color: nth($pair, 1);
    &.is-#{$name} {
      .#{$choices-selector} {
        &__inner {
          border-color: $color;
        }

        &:not(:hover)::after {
          border-color: $color;
        }

        &:focus,
        &.is-focused,
        &:active,
        &.is-active {
          .#{$choices-selector}__inner {
            box-shadow: $input-focus-box-shadow-size bulmaRgba($color, 0.25);
          }
        }

        &:hover,
        &.is-hovered {
          .#{$choices-selector}__inner {
            border-color: bulmaDarken($color, 5%);
          }
        }
      }
    }
  }

  &[required]:not([required='false']),
  &[disabled]:not([disabled='false']) {
    .#{$choices-selector}[data-type*='select-one'] {
      .p6__#{$choices-selector}__button {
        display: none;
      }
    }
  }

  &[disabled]:not([disabled='false']) {
    .#{$choices-selector}[data-type='select-multiple'] {
      .p6__#{$choices-selector}__button {
        display: none;
      }
    }
  }

  .#{$choices-selector} {
    &[data-type*='select-one'] .#{$choices-selector}__inner,
    &[data-type*='select-multiple'] .#{$choices-selector}__inner {
      padding-right: 3.5em;
      padding-bottom: calc(0.5em - 1px);
      padding-left: calc(0.5em - 1px);
      padding-top: calc(0.5em - 1px);
    }

    &[data-type*='select-one'] {
      .p6__choices__button {
        @include cross(#333);

        right: -1.5em;
      }

      .#{$choices-selector}__list--dropdown {
        .#{$choices-selector}__input {
          padding: 0.25em;
        }
      }

      &::after {
        border: 2px solid #000;
        border-radius: 2px;
        border-right: 0;
        border-top: 0;
        content: ' ';
        display: block;
        height: 0.625em;
        margin-top: -0.4375em;
        pointer-events: none;
        position: absolute;
        top: 50%;
        transform: rotate(-45deg);
        transform-origin: center;
        width: 0.625em;
      }

      &.is-open::after {
        margin-top: inherit;
        border-color: inherit;
        transform: rotate(135deg);
        top: calc(50% - 0.125rem);
      }
    }

    &[data-type*='select-multiple'] {
      .#{$choices-selector}__inner {
        padding: 0;
        display: flex;
        align-items: center;
      }

      .#{$choices-selector}__input {
        margin-bottom: 0;
      }

      .#{$choices-selector}__item {
        padding: 3px 5px;
        background-color: $p6-select-multiple-item-bg-color;
        color: $p6-select-multiple-item-text-color;
        border: none;

        .p6__#{$choices-selector}__button {
          @include cross(#000);

          font-size: 0.825em;
          position: relative;
          margin-right: 0.5em;
          margin-left: 0.5em;
        }
      }
    }
  }

  .#{$choices-selector}__inner {
    @include input;

    min-height: initial;
  }

  .#{$choices-selector}__list--single {
    padding: 0;
  }

  .#{$choices-selector}__list--dropdown {
    .#{$choices-selector}__item {
      padding: 0 0 0 0.5em !important;

      &--selectable {
        &.is-highlighted {
          background-color: $p6-select-dropdown-item-highlighted-bg-color;
        }
      }
    }
  }

  .#{$choices-selector}__list--multiple .#{$choices-selector}__item {
    margin: 0.25em;
    font-size: $size-small;

    &.is-highlighted {
      background-color: $p6-select-dropdown-item-highlighted-bg-color;
      border: none;
    }
  }

  .#{$choices-selector}__item {
    position: relative;
  }
}
