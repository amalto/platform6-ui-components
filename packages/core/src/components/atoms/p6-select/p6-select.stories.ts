import { Components } from '../../../components';
import { ComponentProps, getElement, getForm, makeStory, Props } from '../../../shared/storybook';
import { Mode } from '../../../shared/types';

const component = 'p6-select';

export default {
  title: 'Atoms/Select',
  component,
};

const componentProps: ComponentProps = ['name', 'multiple', 'disabled', 'required', 'readOnly', 'placeholder', 'shouldSort', 'disableSearch'];

const getStoryField = (options: HTMLElement[], props?: Props<Components.P6Select>): HTMLElement =>
  getElement(component, options, {
    name: 'language',
    ...props,
  });

const getOption = (value: string, display: string, props?: Props<HTMLOptionElement>): HTMLElement => {
  return getElement('option', display, { value, ...props });
};

export const Default = makeStory<{
  lang: string;
  disabled: boolean;
  multiple: boolean;
  disableSearch: boolean;
  shouldSort: boolean;
  required: boolean;
  readOnly: boolean;
  placeholder: string;
  mode: Mode;
}>({
  componentProps,
  args: {
    lang: 'en',
    disabled: false,
    multiple: false,
    disableSearch: false,
    shouldSort: true,
    required: false,
    readOnly: false,
    placeholder: 'This is a placeholder',
    mode: Mode.default,
  },
  builder: (args): HTMLElement =>
    getStoryField([getOption('no', 'Not selected'), getOption('yes', 'Selected', { selected: true }), getOption('never', 'Disabled', { disabled: true })], args),
});

export const Selected = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField([getOption('no', 'Unselected value'), getOption('yes', 'Selected value', { selected: true })]),
});

export const MultipleSelected = makeStory({
  componentProps,
  builder: (): HTMLElement =>
    getStoryField([getOption('one', 'One', { selected: true }), getOption('two', 'Two', { selected: true }), getOption('three', 'Three')], { multiple: true }),
});

export const Disabled = makeStory<{ disabled: boolean }>({
  componentProps,
  args: { disabled: true },
  builder: (props): HTMLElement => getStoryField([getOption('value', 'Display')], props),
});

export const CustomPlaceholder = makeStory<{ placeholder: string }>({
  componentProps,
  args: {
    placeholder: 'Custom placeholder',
  },
  builder: ({ placeholder }): HTMLElement => getStoryField([getOption('value', 'Display')], { placeholder }),
});

export const Placeholder = makeStory({
  componentProps,
  builder: (): HTMLElement => getStoryField([getOption('', 'This is the placeholder'), getOption('one', 'One'), getOption('two', 'Two'), getOption('three', 'Three')]),
});

export const WithoutSearch = makeStory<{ disableSearch: boolean }>({
  componentProps,
  args: {
    disableSearch: true,
  },
  builder: (props): HTMLElement => getStoryField([getOption('one', 'One'), getOption('two', 'Two'), getOption('three', 'Three')], props),
});

export const Language = makeStory<{ lang: string }>({
  componentProps,
  args: {
    lang: 'en',
  },
  builder: (props): HTMLElement => getStoryField([getOption('one', 'One'), getOption('two', 'Two'), getOption('three', 'Three')], props),
});

export const Form = makeStory<{
  disabled: boolean;
  multiple: boolean;
  required: boolean;
  readOnly: boolean;
}>({
  componentProps,
  args: {
    disabled: false,
    multiple: false,
    required: false,
    readOnly: false,
  },
  builder: (props): HTMLElement =>
    getForm(
      getStoryField(
        [getOption('', 'Select language'), getElement('optgroup', [getOption('fr', 'Fran√ßais'), getOption('en', 'English', { selected: true })], { label: 'Europe' })],
        { name: 'field', ...props },
      ),
    ),
});
