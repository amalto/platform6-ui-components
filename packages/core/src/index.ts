export { Components, JSX } from './components';
export { P6ButtonType } from './components/atoms/p6-button/p6-button';
export { P6CalendarType } from './components/atoms/p6-calendar/p6-calendar';
export { P6CheckboxValue } from './components/atoms/p6-checkbox/p6-checkbox';
export { P6InputValue } from './components/atoms/p6-input/p6-input';
export { P6InputType } from './components/atoms/p6-input/types';
export { Target } from './components/atoms/p6-link/p6-link';
export { P6RadioValue } from './components/atoms/p6-radio/p6-radio';
export { P6SelectControl, P6SelectValue } from './components/atoms/p6-select-native/p6-select-native';
export { P6SwitchValue } from './components/atoms/p6-switch/p6-switch';
export { P6TextareaValue } from './components/atoms/p6-textarea/p6-textarea';
export { P6LanguageValue } from './components/molecules/p6-language/p6-language';
export { P6TranslationControl, P6TranslationValue } from './components/molecules/p6-translation/p6-translation';
export {
  AlignColumnDetail,
  CellValueChangedDetail,
  Column,
  ColumnDefinition,
  ColumnId,
  DataItem,
  EditingCellDetail,
  FilterRowsDetail,
  MoveColumnDetail,
  ResetDefinitionsDetail,
  ResizeColumnDetail,
  Row,
  RowId,
  ShowColumnDetail,
  ShowOptionsDetail,
  SortColumnDetail,
} from './components/organisms/p6-grid/core/entities';
export { ServiceStatus } from './components/templates/p6-service-panel/service-status.enum';
export { P6Control } from './shared/form/control';
export { InvalidEventDetail, isCustomEvent, isInvalidEvent, isValidEvent, ValidEventDetail } from './shared/form/event';
export { Alignment, Mode, modes, Position, positions, Size, sizes, SortOrder } from './shared/types';
export { EnumArrayEntry, enumArrayToObject, enumToArray, EnumType } from './utils/enum';
export { LanguageCode } from './utils/language';
