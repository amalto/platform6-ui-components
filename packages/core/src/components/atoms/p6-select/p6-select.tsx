import { Component, ComponentInterface, Element, h, Host, JSX, Method, Prop, State } from '@stencil/core';
import Choices from 'choices.js';
import { defaultValidationMessage } from '../../../shared/form/validation';
import { isHTMLOptionElement, toArray } from '../../../utils';
import { isEmpty } from '../../../utils/attribute';
import { getL10n, L10n } from '../../../utils/translations';

type ChoiceEventDetail = { value: string };

@Component({
  tag: 'p6-select',
  styleUrl: 'p6-select.scss',
  assetsDirs: ['locales'],
})
export class P6Select implements ComponentInterface {
  /**
   * The name of the select
   */
  @Prop() public name!: string;

  /**
   * Marks the select as multiple
   */
  @Prop() public multiple = false;

  /**
   * The select is not available for interaction. The value will not be submitted with the form
   */
  @Prop() public disabled = false;

  /**
   * Marks the select as required. It can't be submitted without a value
   */
  @Prop() public required = false;

  /**
   * Marks the select as read only.
   */
  @Prop({ attribute: 'readOnly' }) public readOnly = false;

  /**
   * The value of the placeholder to display on the search
   */
  @Prop() public placeholder: string | undefined;

  /**
   * Sort the options by alphabetic order
   */
  @Prop({ attribute: 'shouldSort' }) public shouldSort = false;

  /**
   * Disable the search on the select
   */
  @Prop({ attribute: 'disableSearch' }) public disableSearch = false;

  @Element() private readonly host!: HTMLP6SelectElement;

  @State() private hasError = false;

  private choice: Choices | null = null;

  private nativeSelect: HTMLSelectElement | undefined;

  private l10n: L10n | undefined;

  private selectedValues: string[] = [];

  /**
   * Returns the error message that would be displayed if the user submits the form, or an empty string if no error message.
   * It also triggers the standard error message, such as "this is a required field".
   */
  @Method()
  async validationMessage(): Promise<string> {
    return defaultValidationMessage(this.nativeSelect);
  }

  /**
   * Returns whether a form will validate when it is submitted, without having to submit it.
   */
  @Method()
  async checkValidity(): Promise<boolean> {
    return Promise.resolve(this.nativeSelect?.checkValidity() || true);
  }

  private getSelectedValueIndex(value: string): number {
    return this.selectedValues.indexOf(value);
  }

  async componentWillLoad(): Promise<void> {
    this.l10n = await getL10n(this.host);
    this.host.addEventListener('focusout', this.internalCheckValidity.bind(this));

    this.host.addEventListener('addItem', event => {
      const customEvent = event as CustomEvent<ChoiceEventDetail>;
      const { value } = customEvent.detail;
      if (this.getSelectedValueIndex(value) !== -1) {
        return;
      }
      this.selectedValues.push(value);
      this.internalCheckValidity();
    });

    this.host.addEventListener('removeItem', event => {
      const customEvent = event as CustomEvent<ChoiceEventDetail>;
      const idx = this.getSelectedValueIndex(customEvent.detail.value);
      if (idx !== -1) {
        this.selectedValues.splice(idx, 1);
      }
      this.internalCheckValidity();
    });

    // Default values
    this.selectedValues = toArray(this.host.children)
      .filter(isHTMLOptionElement)
      .filter(child => child.selected)
      .map(option => option.value);
  }

  componentWillRender(): Promise<void> | void {
    // destroy choices element to restore previous dom structure
    // so vdom can replace the element correctly
    this.destroy();
  }

  private getOptionChildren(): HTMLOptionElement[] {
    return toArray(this.host.children[0].children).filter(isHTMLOptionElement);
  }

  render(): JSX.Element {
    const visibleSelectAttributes = {
      name: this.readOnly ? '' : this.name,
      disabled: this.disabled || this.readOnly,
      multiple: this.multiple,
      required: this.required,
      class: { 'is-danger': true },
    };

    const hiddenSelectAttributes = {
      name: this.readOnly ? this.name : '',
      disabled: !this.readOnly,
      multiple: this.multiple,
      style: { display: 'none' },
    };

    return (
      <Host
        aria-disabled={this.disabled ? 'true' : null}
        class={{
          'is-danger': this.hasError,
        }}
        required={this.required}
      >
        <select {...hiddenSelectAttributes}>
          <slot />
        </select>
        <select
          ref={(input): void => {
            this.nativeSelect = input;
          }}
          {...visibleSelectAttributes}
        />
      </Host>
    );
  }

  componentDidRender(): void {
    this.init();
  }

  componentDidLoad(): void {
    if (this.selectedValues.length > 0) {
      this.internalCheckValidity();
    }
  }

  disconnectedCallback(): void {
    this.destroy();
  }

  private init(): void {
    // cf. https://github.com/jshjohnson/Choices
    const props = {
      // Usage: Optionally suppress console errors and warnings.
      silent: true,
      // Usage: Whether a user can remove items.
      removeItems: true,
      // Usage: Whether each item should have a remove button.
      removeItemButton: true,
      // Usage: Whether a search area should be shown. Note: Multiple select boxes will always show search areas.
      searchEnabled: !this.disableSearch,
      // Usage: Whether choices and groups should be sorted. If false, choices/groups will appear in the order they were given.
      shouldSort: this.shouldSort,
      // Usage: The value of the search inputs placeholder.
      placeholder: !isEmpty(this.placeholder),
      searchPlaceholderValue: this.placeholder,
      // Usage: The text that is shown whilst choices are being populated via AJAX.
      loadingText: this.l10n?.loadingText,
      // Usage: The text that is shown when a user's search has returned no results. Optionally pass a function returning a string.
      noResultsText: this.l10n?.noResultsText,
      // Usage: The text that is shown when a user has selected all possible choices. Optionally pass a function returning a string.
      noChoicesText: this.l10n?.noChoicesText,
      // Usage: The text that is shown when a user hovers over a selectable choice.
      itemSelectText: '',
      classNames: {
        containerOuter: 'choices',
        containerInner: 'choices__inner',
        input: 'choices__input',
        inputCloned: 'choices__input--cloned',
        list: 'choices__list',
        listItems: 'choices__list--multiple',
        listSingle: 'choices__list--single',
        listDropdown: 'choices__list--dropdown',
        item: 'choices__item',
        itemSelectable: 'choices__item--selectable',
        itemDisabled: 'choices__item--disabled',
        itemChoice: 'choices__item--choice',
        placeholder: 'choices__placeholder',
        group: 'choices__group',
        groupHeading: 'choices__heading',
        button: 'p6__choices__button',
        activeState: 'is-active',
        focusState: 'is-focused',
        openState: 'is-open',
        disabledState: 'is-disabled',
        highlightedState: 'is-highlighted',
        selectedState: 'is-selected',
        flippedState: 'is-flipped',
        loadingState: 'is-loading',
        noResults: 'has-no-results',
        noChoices: 'has-no-choices',
      },
      choices: this.getOptionChildren().map(option => {
        return {
          value: option.value,
          label: option.label,
          disabled: option.disabled,
          selected: this.selectedValues.indexOf(option.value) !== -1,
        };
      }),
    };

    if (this.nativeSelect !== undefined) {
      this.choice = new Choices(this.nativeSelect, props);
    }
  }

  private destroy(): void {
    if (this.choice !== null) {
      this.choice.destroy();
      this.choice = null;
    }
  }

  private async internalCheckValidity(): Promise<void> {
    await this.checkValidity();
    const msg = await this.validationMessage();
    this.hasError = msg !== '';
  }
}
