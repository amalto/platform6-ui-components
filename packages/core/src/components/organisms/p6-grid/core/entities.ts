import { EventEmitter, JSX } from '@stencil/core';
import { Alignment, Direction, SortOrder } from '../../../../shared/types';

export type ColumnId = string;

export type DataItem = Record<string, unknown>;

export interface ColumnDefinition<T extends DataItem> {
  id?: ColumnId;
  label: string;
  field?: string | keyof T;
  editable?: boolean;
  filtreable?: boolean;
  sortable?: boolean;
  hidden?: boolean;
  width?: number;
  align?: Alignment;
  color?: string;
  sortOrder?: SortOrder;
  disableHeaderMenu?: boolean;
  getValue?: (data: T, context: ColumnDefinition<T>) => string;
  setValue?: (value: string, data: T, context: ColumnDefinition<T>) => T;
  cellEditor?: (data: T, context: ColumnDefinition<T>) => JSX.Element & CellEditor<T>;
  cellRenderer?: (data: T, rowId: RowId, context: ColumnDefinition<T>) => string | HTMLElement;
}

export interface Column<T extends DataItem> extends ColumnDefinition<T> {
  id: ColumnId;
  editable: boolean;
  hidden: boolean;
  width: number;
  align: Alignment;
  sortOrder: SortOrder;
  disableHeaderMenu: boolean;
  getValue: (data: T, context: ColumnDefinition<T>) => string;
  setValue: (value: string, data: T, context: ColumnDefinition<T>) => T;
}

export interface CellEditor<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  value: string;

  p6GridCellValueChanged: EventEmitter<CellValueChangedDetail<T>>;
  p6GridEditingCell: EventEmitter<EditingCellDetail<T>>;
}
export type RowId = string;

export interface Row<T extends DataItem> {
  id: RowId;
  data: T;
}

interface ColumnEventDetail<T extends DataItem> {
  column: Column<T>;
}

export interface MoveColumnDetail<T extends DataItem> extends ColumnEventDetail<T> {
  direction: Direction;
}
export type ResizeColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type AlignColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type SortColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export type ShowColumnDetail<T extends DataItem> = ColumnEventDetail<T>;

export interface CellValueChangedDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  oldValue: string;
}

export enum EditingCellStatus {
  Start,
  Stop,
  Cancel,
}

export interface EditingCellDetail<T extends DataItem> {
  column: Column<T>;
  row: Row<T>;
  status: EditingCellStatus;
}

export interface ResetDefinitionsDetail {
  reset: boolean;
}

export interface ShowOptionsDetail {
  visible: boolean;
}

export interface FilterRowsDetail {
  value: string;
}
