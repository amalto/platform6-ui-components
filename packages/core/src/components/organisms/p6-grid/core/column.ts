import { Alignment, Direction, SortOrder } from '../../../../shared/types';
import { throwError } from '../../../../utils/error';
import { Column, ColumnDefinition, ColumnId, DataItem } from './entities';

export const MIN_WIDTH = 75;
export const DEFAULT_WIDTH = MIN_WIDTH;
export const INC_WIDTH = 10;

export function defaultGetValue<T extends DataItem>(data: T, context: ColumnDefinition<T>): string {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'getValue' function for ${JSON.stringify(context)}`);
  }

  return `${data[context.field as keyof T]}`;
}

export function defaultSetValue<T extends DataItem>(value: string, data: T, context: ColumnDefinition<T>): T {
  if (context.field === undefined || data[context.field] === undefined) {
    return throwError(`field is undefined or doesn't exist, you should provide a 'setValue' function for ${JSON.stringify(context)}`);
  }

  return {
    ...data,
    [context.field as keyof T]: value,
  };
}

let columnIds = 0;
export function fromDefinition<T extends DataItem>(column: ColumnDefinition<T>): Column<T> {
  // eslint-disable-next-line no-plusplus
  const id = column.id || column.field || columnIds++;

  return {
    ...column,
    id: `col-${id}`,
    editable: column.editable || false,
    hidden: column.hidden || false,
    width: column.width || 100,
    align: column.align || Alignment.center,
    sortOrder: column.sortOrder || SortOrder.none,
    disableHeaderMenu: column.disableHeaderMenu || false,
    getValue: column.getValue || defaultGetValue,
    setValue: column.setValue || defaultSetValue,
  };
}

export function isHidden<T extends DataItem>(column: Column<T>): boolean {
  return !!column.hidden;
}

export function replaceColumn<T extends DataItem>(updatedColumn: Column<T>, allColumns: Column<T>[]): Column<T>[] {
  return allColumns.map(col => (col.id === updatedColumn.id ? updatedColumn : col));
}

export function move<T extends DataItem>(id: ColumnId, direction: Direction, columns: Column<T>[]): Column<T>[] {
  const from = columns.findIndex(col => col.id === id);
  const to = from + (direction === Direction.left ? -1 : 1);

  if (from < 0 || to < 0 || to >= columns.length) {
    return columns;
  }

  const result = columns.slice();
  const item = result.splice(from, 1);

  return result.slice(0, to).concat(item).concat(result.slice(to, columns.length));
}
