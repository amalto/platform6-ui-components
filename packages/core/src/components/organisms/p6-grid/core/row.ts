import { SortOrder } from '../../../../shared/types';
import { Column, DataItem, Row, RowId } from './entities';

let rowIds = 0;
export function fromData<T extends DataItem>(data: T): Row<T> {
  rowIds += 1;

  return {
    id: `row-${rowIds}`,
    data,
  };
}

export function compareRow<T extends DataItem>(column: Column<T> | undefined): (a: Row<T>, b: Row<T>) => number {
  const sortOrder = column?.sortOrder || SortOrder.none;
  const factor = sortOrder === SortOrder.desc ? -1 : 1;

  return (a: Row<T>, b: Row<T>): number => {
    if (column === undefined || column.sortable === false || sortOrder === SortOrder.none) {
      return 0;
    }

    const valueA = column.getValue(a.data, column);
    const valueB = column.getValue(b.data, column);

    return valueA.localeCompare(valueB) * factor;
  };
}

export function filterBy<T extends DataItem>(searchValue: string, columns: Column<T>[]): (row: Row<T>) => boolean {
  return row => columns.filter(col => col.filtreable !== false).some(col => col.getValue(row.data, col).includes(searchValue));
}

export function rangeSelectRow<T extends DataItem>(currentRowIdSelect: RowId, selectedRows: Set<RowId>, displayedRow: Row<T>[]): Set<RowId> {
  const currentSelectedIndex = displayedRow.findIndex(row => row.id === currentRowIdSelect);

  if (currentSelectedIndex === -1) {
    return selectedRows;
  }

  const previousRowIdSelected = Array.from(selectedRows)
    .reverse()
    .find(rowId => displayedRow.findIndex(row => row.id === rowId) > -1);

  if (previousRowIdSelected === undefined) {
    return new Set(selectedRows.add(currentRowIdSelect));
  }

  const previousSelectedIndex = displayedRow.findIndex(row => row.id === previousRowIdSelected);

  let [from, to] = [previousSelectedIndex, currentSelectedIndex];
  let excludedFromSlice = currentRowIdSelect;
  if (currentSelectedIndex < previousSelectedIndex) {
    [from, to] = [to, from];
    excludedFromSlice = previousRowIdSelected;
  }

  return displayedRow
    .slice(from, to)
    .reduce((selected, cur) => {
      return cur.id === undefined ? selected : selected.add(cur.id);
    }, new Set(selectedRows))
    .add(excludedFromSlice);
}

export function replaceRow<T extends DataItem>(updatedRow: Row<T>, allRows: Row<T>[]): Row<T>[] {
  return allRows.map(row => (row.id === updatedRow.id ? updatedRow : row));
}
