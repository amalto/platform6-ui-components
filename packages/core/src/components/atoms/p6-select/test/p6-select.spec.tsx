import { newSpecPage } from '@stencil/core/testing';
import { P6Select } from '../p6-select';

describe('p6-select', () => {
  it('renders', async () => {
    const page = await newSpecPage({
      components: [P6Select],
      html: `<p6-select><option>value</option></p6-select>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('multiple', async () => {
    const page = await newSpecPage({
      components: [P6Select],
      html: `<p6-select multiple><option>value</option></p6-select>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('disabled', async () => {
    const page = await newSpecPage({
      components: [P6Select],
      html: `<p6-select disabled><option>value</option></p6-select>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
