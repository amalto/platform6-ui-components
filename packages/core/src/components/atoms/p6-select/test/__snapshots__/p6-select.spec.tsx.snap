// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`p6-select disabled 1`] = `
<p6-select aria-disabled="true" disabled="">
  <!---->
  <select disabled="" style="display: none;">
    <option>
      value
    </option>
  </select>
  <select class="is-danger" disabled=""></select>
</p6-select>
`;

exports[`p6-select multiple 1`] = `
<p6-select multiple="">
  <!---->
  <select disabled="" multiple="" style="display: none;">
    <option>
      value
    </option>
  </select>
  <select class="is-danger" multiple=""></select>
</p6-select>
`;

exports[`p6-select renders 1`] = `
<p6-select>
  <!---->
  <select disabled="" style="display: none;">
    <option>
      value
    </option>
  </select>
  <select class="is-danger"></select>
</p6-select>
`;
